src/psyclone/tests/psyir/frontend/where_handler_test.py

```diff
@@ -0,0 +1,345 @@
+                                 "END WHERE\n")
+    fparser2spec = Fortran2003.Where_Construct(reader)
+    processor.process_nodes(fake_parent, [fparser2spec], None)
+    assert isinstance(fake_parent.children[0], CodeBlock)
+
+
+def test_missing_array_notation_lhs(parser):
+    ''' Check that we get a code block if the WHERE does not use explicit
+    array syntax on the LHS of an assignment within the body. '''
+    fake_parent = Schedule()
+    processor = Fparser2Reader()
+    reader = FortranStringReader("WHERE (ptsu(:,:,:) /= 0._wp)\n"
+                                 "  z1_st = 1._wp / ptsu(:, :, :)\n"
+                                 "END WHERE\n")
+    fparser2spec = Fortran2003.Where_Construct(reader)
+    processor.process_nodes(fake_parent, [fparser2spec], None)
+    assert isinstance(fake_parent.children[0], CodeBlock)
+
+
+def test_where_array_notation_rank(parser):
+    ''' Test that the _array_notation_rank() utility raises the expected
+    errors when passed an unsupported Array object. '''
+    from psyclone.psyGen import Array, Literal
+    my_array = Array("my_array", None)
+    processor = Fparser2Reader()
+    with pytest.raises(NotImplementedError) as err:
+        processor._array_notation_rank(my_array)
+    assert ("Array reference in the PSyIR must have at least one child but "
+            "'my_array'" in str(err.value))
+    # Give the Array one child that is not a CodeBlock
+    my_array.addchild(Literal("2", my_array))
+    with pytest.raises(NotImplementedError) as err:
+        processor._array_notation_rank(my_array)
+    assert ("that uses Fortran array notation is assumed to have at least "
+            "one CodeBlock as its child " in str(err.value))
+    my_array._children = []
+    my_array.addchild(CodeBlock([Fortran2003.Literal_Constant("1")],
+                                CodeBlock.Structure.EXPRESSION, my_array))
+    with pytest.raises(NotImplementedError) as err:
+        processor._array_notation_rank(my_array)
+    assert ("Only array notation of the form my_array(:, :, ...) is supported"
+            in str(err.value))
+
+
+def test_different_ranks_error(parser):
+    ''' Check that a WHERE construct containing array references of different
+    ranks results in the creation of a CodeBlock. '''
+    fake_parent = Schedule()
+    processor = Fparser2Reader()
+    # Invalid Fortran
+    reader = FortranStringReader("WHERE (dry(:, :, :))\n"
+                                 "  z1_st(:, :) = depth / ptsu(:, :, :)\n"
+                                 "END WHERE\n")
+    fparser2spec = Fortran2003.Where_Construct(reader)
+    # We should end up with a CodeBlock in the PSyIR
+    processor.process_nodes(fake_parent, [fparser2spec], None)
+    assert isinstance(fake_parent.children[0], CodeBlock)
+
+
+def test_array_notation_rank(parser):
+    ''' Check that the _array_notation_rank() utility handles various examples
+    of array notation. '''
+    from psyclone.psyGen import Assignment, Loop
```

Comment:
W:140, 4: Unused variable 'Assignment' (unused-variable)
W:140, 4: Unused variable 'Loop' (unused-variable)

Reply:
I've rationalised the imports.